### PR TITLE
[Train] Torch Prepare utilities

### PR DESCRIPTION
**Pull request by amogkam**
## Checks

- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: python/ray/train/backends/torch.py [hunk 1]
```diff
@@ -144,3 +145,144 @@ def on_shutdown(self, worker_group: WorkerGroup,
 
         worker_group.execute(
             shutdown_torch, destroy_process_group=len(worker_group) > 1)
+
+
+if torch:
```
Comment (matthewdeng) [inline python/ray/train/backends/torch.py:150]: Hmm is there a cleaner way to do this without wrapping all the functions?

File: python/ray/train/backends/torch.py [hunk 2]
```diff
@@ -144,3 +145,144 @@ def on_shutdown(self, worker_group: WorkerGroup,
...
+    from torch.nn.parallel import DistributedDataParallel
+    from torch.utils.data import DistributedSampler, DataLoader, \
+        IterableDataset
+
+    def _get_torch_device() -> torch.device:
```
Comment (matthewdeng) [inline python/ray/train/backends/torch.py:155]: Can we expose this for users?

File: python/ray/train/backends/torch.py [hunk 3]
```diff
@@ -144,3 +145,144 @@ def on_shutdown(self, worker_group: WorkerGroup,
...
+
+        def __iter__(self):
+            iterator = iter(self._dataloader)
+            if self._device is None:
+                yield from iterator
+
+            for item in iterator:
+                yield (i.to(self.device) for i in item)
+
+    def prepare(
```
Comment (matthewdeng) [inline python/ray/train/backends/torch.py:187]: Name this `prepare_module`/`prepare_model`?

File: python/ray/train/examples/train_linear_dataset_example.py [hunk 4]
```diff
@@ -122,6 +113,13 @@ def train_func(config):
             feature_column_dtypes=[torch.float],
             batch_size=batch_size)
 
+        rank = train.local_rank()
+
+        if torch.cuda.is_available():
+            device = torch.device("cpu")
+        else:
+            device = torch.device(f"cuda:{rank}")
```
Comment (matthewdeng) [inline python/ray/train/examples/train_linear_dataset_example.py:121]: Example for above comment, looks like this is reversed 😄

File: doc/source/train/train.rst [hunk 5]
```diff
@@ -85,10 +85,10 @@ system. Let's take following simple examples:
 
     Now let's convert this to a distributed multi-worker training function!
 
-    First, update the training function code to use PyTorch's
-    ``DistributedDataParallel``. With Ray Train, you just pass in your distributed
-    data parallel code as as you would normally run it with
-    ``torch.distributed.launch``.
+    All you have to do is use the ``ray.train.torch.prepare`` utility function to
+    easily setup your model, data, optimizer, & loss for distributed training.
```
Comment (matthewdeng) [inline doc/source/train/train.rst]: This mentions optimizer and loss but isn't implemented (yet?).

File: python/ray/train/examples/train_linear_dataset_example.py (hunk 4 above)
Comment (matthewdeng) [reply python/ray/train/examples/train_linear_dataset_example.py:121]: Also wondering if there's a way to encapsulate this into the utilities, so that the user doesn't need to manually handle the device anywhere.

File: python/ray/train/backends/__init__.py [hunk 6]
```diff
@@ -1,7 +1,9 @@
 from ray.train.backends.backend import BackendConfig
 from ray.train.backends.horovod import HorovodConfig
 from ray.train.backends.tensorflow import TensorflowConfig
-from ray.train.backends.torch import TorchConfig
+
+# Backwards Compatiblity
+from ray.train.torch import TorchConfig
```
Comment (matthewdeng) [inline python/ray/train/backends/__init__.py:6]: Is this change needed? Doesn't `ray.train.torch` just point back to `ray.train.backends.torch`?

Comment (matthewdeng) [reply python/ray/train/backends/__init__.py:6]: Can you also explain what the long term plan for this is? Maybe remove `backends` package entirely?

File: python/ray/train/session.py [hunk 7]
```diff
@@ -413,3 +415,23 @@ def train_func():
...
+def world_size() -> int:
+    """Get the current world size for this training execution.
+
+    .. code-block:: python
+
+        import time
+        from ray import sgd
+
+        def train_func():
+            assert sgd.world_size() == 4
```
Comment (matthewdeng) [inline python/ray/train/session.py:429]: sgd -> train

File: python/ray/train/backends/__init__.py (hunk 6 above)
Comment (amogkam) [reply python/ray/train/backends/__init__.py:6]: The idea is that from a user-facing API perspective, it's better to do `model = train.torch.prepare(model)` instead of `model = train.backends.torch.prepare(model)`.

I think ideally yes, we can remove the `backends` package entirely.

File: python/ray/train/examples/train_linear_dataset_example.py (hunk 4 above)
Comment (matthewdeng) [reply python/ray/train/examples/train_linear_dataset_example.py:121]: For this example we're using the `device` to move the tensors, but I think the `_WrappedDataLoader` handles this?

File: doc/source/train/train.rst (hunk 5 above)
Comment (amogkam) [reply doc/source/train/train.rst]: That's right not implemented yet. The description has been updated.

File: python/ray/train/examples/train_linear_dataset_example.py (hunk 4 above)
Comment (amogkam) [reply python/ray/train/examples/train_linear_dataset_example.py:121]: Here we're not using `prepare_data_loader` or a `DataLoader` at all since it's coming from a Ray Dataset shard.

File: python/ray/train/backends/torch.py (hunk 3 above)
Comment (amogkam) [reply python/ray/train/backends/torch.py:187]: I had this as an arbitrary `prepare` to support optimizer preparation, etc. in the future.

File: python/ray/train/backends/torch.py (hunk 2 above)
Comment (amogkam) [reply python/ray/train/backends/torch.py:155]: Yep good point!

File: python/ray/train/examples/train_linear_dataset_example.py (hunk 4 above)
Comment (amogkam) [reply python/ray/train/examples/train_linear_dataset_example.py:121]: Not sure what the best way to handle Ray Dataset/to_torch usability would be, but left that as out of scope for this PR

File: python/ray/train/session.py (hunk 7 above)
Comment (amogkam) [reply python/ray/train/session.py]: fixed!

File: python/ray/train/backends/torch.py (hunk 1 above)
Comment (amogkam) [reply python/ray/train/backends/torch.py:150]: One option is to just assume that torch is installed in torch.py, but move `TorchConfig` to backend.py

Comment (amogkam) [reply python/ray/train/backends/torch.py:150]: Actually, maybe we don't even need backwards compatibility..after all this is still alpha. Just move all the files in `backends` package and just have a dummy `backends.py` that raises an error saying all the stuff has been moved.

File: doc/source/train/user_guide.rst [hunk 8]
```diff
@@ -64,59 +64,25 @@ training.
...
             # Create model.
             model = NeuralNetwork()
-            model = model.to(device)
-            model = DistributedDataParallel(
-                model,
-                device_ids=[train.local_rank()] if torch.cuda.is_available() else None)
+
+            # Use the ``prepare`` function to automatically move your model to the right device and wrap it in
+            # ``DistributedDataParallel``
+            model = train.torch.prepare(model)
```
Comment (amogkam) [inline doc/source/train/user_guide.rst:82]: TODO: Figure out a better code snippet to use here

File: python/ray/train/backends/__init__.py (hunk 6 above)
Comment (matthewdeng) [reply python/ray/train/backends/__init__.py:6]: Oh yeah I definitely agree with that, the original comment was just about how `TorchConfig` is defined in `python/ray/train/backends/torch.py` so this import is `ray.train.backend.torch` -> `ray.train.torch.` -> `ray.train.backend.torch`.

File: doc/source/train/user_guide.rst (hunk 8 above)
Comment (richardliaw) [reply doc/source/train/user_guide.rst:82]: nice!

File: python/ray/train/backends/torch.py (hunk 3 above)
Comment (matthewdeng) [reply python/ray/train/backends/torch.py:188]: Hmm I'm a little confused of what the end state is here. Are we still planning to do something like: `prepare (*args, TorchPrepareConfig)`? Why is `prepare_dataloader` separate?

File: python/ray/train/examples/train_linear_dataset_example.py (hunk 4 above)
Comment (matthewdeng) [reply python/ray/train/examples/train_linear_dataset_example.py]: Gotcha that makes sense!

**Comment by matthewdeng** [comment]
Really like how this is coming together, the examples and docs are noticeably simpler! For the sake of notekeeping on intentionality, could you update either the PR description or docs to describe:

1. What are we solving in this PR?
2. What are we _not_ solving in the PR?
     1. Which of these are TODOs that will be addressed in the future?

File: python/ray/train/backends/torch.py (hunk 1 above)
Comment (amogkam) [reply python/ray/train/backends/torch.py]: Did the refactoring in a separate PR: https://github.com/ray-project/ray/pull/20312

File: python/ray/train/backends/torch.py (hunk 3 above)
Comment (amogkam) [reply python/ray/train/backends/torch.py]: Agreed offline- changed this to `prepare_model`

File: python/ray/train/backends/__init__.py (hunk 6 above)
Comment (amogkam) [reply python/ray/train/backends/__init__.py]: Did the refactoring in a separate PR! https://github.com/ray-project/ray/pull/20312

File: python/ray/train/session.py [hunk 9]
```diff
@@ -413,3 +415,23 @@ def train_func():
     """
     session = get_session()
     session.checkpoint(**kwargs)
+
+
+def world_size() -> int:
+    """Get the current world size for this training execution.
```
Comment (matthewdeng) [inline python/ray/train/session.py:421]:
```suggestion
    """Get the current world size (i.e. total number of workers) for this training execution.
```

File: python/ray/train/torch.py [hunk 10]
```diff
@@ -144,3 +139,135 @@ def on_shutdown(self, worker_group: WorkerGroup,
...
+
+
+def get_device() -> torch.device:
+    """Gets the correct torch device to use for training."""
+    rank = train.local_rank()
+
+    if not torch.cuda.is_available():
+        device = torch.device("cpu")
+    else:
+        device = torch.device(f"cuda:{rank}")
```
Comment (matthewdeng) [inline python/ray/train/torch.py:174]: nit for readability
```suggestion
    if torch.cuda.is_available():
        rank = train.local_rank()
        device = torch.device(f"cuda:{rank}")
    else:
        device = torch.device("cpu")
```

File: python/ray/train/torch.py [hunk 11]
```diff
@@ -144,3 +139,135 @@ def on_shutdown(self, worker_group: WorkerGroup,
...
+            the provided DataLoader.
+        move_to_device (bool): If set, automatically move the data
+            returned by the data loader to the correct device.
+    """
+
+    if train.world_size() > 1 \
+        and not isinstance(data_loader.sampler, DistributedSampler) \
+        and not (hasattr(data_loader, "dataset")
+                 and isinstance(data_loader.dataset, IterableDataset)) \
+            and add_dist_sampler:
```
Comment (matthewdeng) [inline python/ray/train/torch.py:246]: Add a comment explaining this?

File: python/ray/train/torch.py [hunk 12]
```diff
@@ -144,3 +139,135 @@ def on_shutdown(self, worker_group: WorkerGroup,
...
+
+        def with_sampler(loader):
+            # Automatically set the DistributedSampler
+
+            shuffle = not isinstance(loader.sampler, SequentialSampler)
```
Comment (matthewdeng) [inline python/ray/train/torch.py:251]: Add a comment explaining this? Also the reason for why we set `shuffle` to False in `data_loader_args` in relation to `DistributedSampler`.

**Review by matthewdeng — APPROVED** [review]

**Review by matthewdeng — COMMENTED** [review]
LGTM (please fix tests).

File: python/ray/train/torch.py (hunk 11 above)
Comment (amogkam) [reply python/ray/train/torch.py:250]: Done!

File: python/ray/train/torch.py (hunk 12 above)
Comment (amogkam) [reply python/ray/train/torch.py:262]: Done!
